### PR TITLE
router: support for request/response header manipulation in weighted clusters

### PR DESCRIPTION
**Pull request by zuercher**
Implements the `request_headers_to_add`, `response_headers_to_add`, and `response_headers_to_remove` fields added to weighted clusters by envoyproxy/data-plane-api#441.

*Risk Level*: Low - no change in behavior without configuration changes

*Testing*: unit and integration tests

*Docs Changes*: envoyproxy/data-plane-api#531

*Release Notes*: updated

*Fixes*: #2455 

Signed-off-by: Stephan Zuercher <stephan@turbinelabs.io>

### CODE REVIEW

**Comment by htuch** [comment]
@junr03 can you take a first pass on this? Thanks.

File: source/common/router/config_impl.cc [hunk 1]
```diff
@@ -536,6 +518,28 @@ void RouteEntryImplBase::validateClusters(Upstream::ClusterManager& cm) const {
...
 
+RouteEntryImplBase::WeightedClusterEntry::WeightedClusterEntry(
+    const RouteEntryImplBase* parent, const std::string runtime_key, Runtime::Loader& loader,
+    const envoy::api::v2::route::WeightedCluster_ClusterWeight& cluster)
+    : DynamicRouteEntry(parent, cluster.name()), runtime_key_(runtime_key), loader_(loader),
+      cluster_weight_(PROTOBUF_GET_WRAPPED_REQUIRED(cluster, weight)),
+      request_headers_parser_(HeaderParser::configure(cluster.request_headers_to_add())),
+      response_headers_parser_(HeaderParser::configure(cluster.response_headers_to_add(),
+                                                       cluster.response_headers_to_remove())) {
+  if (cluster.has_metadata_match()) {
```
Comment (junr03) [inline source/common/router/config_impl.cc]: thanks for the cleanup here!

File: test/integration/header_integration_test.cc [hunk 2]
```diff
@@ -643,28 +683,132 @@ TEST_P(HeaderIntegrationTest, TestRouteConfigVirtualHostAndRouteReplaceHeaderMan
...
+          {":authority", "vhost-headers.com"},
+          {"x-unmodified", "request"},
+          {"x-weighted-cluster-request", "weighted-cluster-1"},
+          {"x-route-request", "route"},
+          {"x-vhost-request", "vhost"},
+          {"x-routeconfig-request", "routeconfig"},
+          {":path", "/vhost-route-and-weighted-clusters"},
+          {":method", "GET"},
+      },
+      Http::TestHeaderMapImpl{
```
Comment (junr03) [inline test/integration/header_integration_test.cc]: I might be missing something in how these tests are run, but why not have the `x-weighted-cluster-response-remove` header here?

Comment (zuercher) [reply test/integration/header_integration_test.cc]: It's in the previous test. You can think of these tests as existing in append/replace pairs, and I decided not to repeat the testing of header removal in the replace tests.

Comment (junr03) [reply test/integration/header_integration_test.cc]: Yep, I was wondering if having the remove behavior tested in both places would ensure that remove behavior works correctly regardless of adding scheme (append/replace). But if you don't think that gives us anything I am good with it.

**Comment by junr03** [comment]
@htuch lgtm, do you mind taking a final pass?

**Pull request merged**